certificates/src/lib.rs

```diff
@@ -125,3 +186,363 @@ impl CertificateValidator {
+        dbg!(&ext_key_usage);
+        // Per
+        // https://man.openbsd.org/X509_check_purpose.3#X509_PURPOSE_SSL_SERVER,
+        // if we have the Extended Key Usage extension, we must have either the
+        // server auth bit or one of a couple proprietary (Netscape/Microsoft)
+        // bits; we'll ignore the proprietary bits and require server auth.
+        //
+        // TODO-correctness: Is it right to also accept "any" here?
+        if !ext_key_usage.value.any && !ext_key_usage.value.server_auth {
+            return Err(CertificateError::UnsupportedPurpose);
+        }
+    }
+
+    // We either don't have Key Usage / Extended Key Usage, or we do but they're
+    // satisfactory.
+    Ok(())
+}
+
+fn cert_dns_names(cert: &X509) -> Result<Vec<DnsName>, CertificateError> {
+    let mut dns_names = Vec::new();
+
+    // Per RFC 2818 § 3.1:
+    //
+    // > If a subjectAltName extension of type dNSName is present, that MUST
+    // > be used as the identity. Otherwise, the (most specific) Common Name
+    // > field in the Subject field of the certificate MUST be used.
+    //
+    // we first check for any dnsname entries in `cert.subject_alt_names()`, and
+    // failing that we use the common name from the subject field.
+
+    if let Some(sans) = cert.subject_alt_names() {
+        for san in sans {
+            if let Some(dnsname) = san.dnsname() {
+                dns_names.push(
+                    DnsName::from_utf8(dnsname)
+                        .map_err(CertificateError::InvalidDnsName)?,
+                );
+            }
+        }
+    }
+
+    // If we found at least one dNSName in the SANs; we're done.
+    if !dns_names.is_empty() {
+        return Ok(dns_names);
+    }
+
+    let sn = cert.subject_name();
+
+    for cn in sn.entries_by_nid(Nid::COMMONNAME) {
+        let cn =
+            cn.data().as_utf8().map_err(CertificateError::NonUtf8CommonName)?;
+        dns_names.push(
+            DnsName::from_utf8(cn)
+                .map_err(CertificateError::InvalidCommonNameAsDnsName)?,
+        );
+    }
+
+    // Per the RFC, we should now choose the "most specific" common name, but
+    //
+    // (a) it doesn't say how to do that, and
+    // (b) we don't really expect to get here (we expect to find SANs)
+    //
+    // so we'll bail out if we found multiple common names. If we only found
+    // one, picking the most specific is easy!
+    //
+    // The crate we use for generating test cert parameters (rcgen) does not
+    // even support multiple common name entries.
+    match dns_names.len() {
+        0 => Err(CertificateError::NoCommonName),
+        1 => Ok(dns_names),
+        _ => Err(CertificateError::NoDnsNameSansMultipleCommonNames),
+    }
+}
+
+// Returns `Ok(true)` if `dns_name` (from a cert) matches `hostname`, including
+// handling a leading `*` wildcard:
+//
+// * If `hostname` begins with a wildcard, `dns_name` must be an exact match
```

Comment:
I read `hostname` here like "an actual DNS name, as might be provided when someone makes an HTTP request to Nexus".  I wouldn't expect it to be possible for that to have wildcards.  Is it really "a hostname pattern", sort of of our own invention, so that we can validate `*.sys.$delegated_domain`?

Reply:
That is how I used it from wicket (to check that the cert uploaded would work for `*.sys.$delegated_domain`, exactly). If wicketd should only be checking for `recovery.sys.$delegated_domain`, I could change this to only match actual hostnames?